### PR TITLE
HBASE-22699 refactor isMetaClearingException

### PR DESCRIPTION
**Pull request by johnhomsea**

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m  6s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ branch-2 Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 15s |  Maven dependency ordering for branch  |
| :green_heart: |  mvninstall  |   5m 17s |  branch-2 passed  |
| :green_heart: |  compile  |   1m 22s |  branch-2 passed  |
| :green_heart: |  checkstyle  |   2m  4s |  branch-2 passed  |
| :green_heart: |  shadedjars  |   4m 39s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   1m  5s |  branch-2 passed  |
| :blue_heart: |  spotbugs  |   3m 26s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 43s |  branch-2 passed  |
||| _ Patch Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 15s |  Maven dependency ordering for patch  |
| :green_heart: |  mvninstall  |   4m 48s |  the patch passed  |
| :green_heart: |  compile  |   1m 23s |  the patch passed  |
| :green_heart: |  javac  |   1m 23s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 51s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m  8s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  15m  5s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 59s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 37s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  |   3m 37s |  hbase-client in the patch failed.  |
| :broken_heart: |  unit  | 292m 51s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   1m  5s |  The patch does not generate ASF License warnings.  |
|  |   | 357m 28s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestAsyncProcess |
|   | hadoop.hbase.client.replication.TestReplicationAdminWithClusters |
|   | hadoop.hbase.client.TestFromClientSide3 |
|   | hadoop.hbase.client.TestSnapshotDFSTemporaryDirectory |
|   | hadoop.hbase.client.TestAsyncTableAdminApi |
|   | hadoop.hbase.replication.TestReplicationSmallTests |
|   | hadoop.hbase.client.TestCloneSnapshotFromClientNormal |
|   | hadoop.hbase.client.TestFromClientSideWithCoprocessor |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-578/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/578 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 2b2425832026 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-578/out/precommit/personality/provided.sh |
| git revision | branch-2 / aeacfd3d59 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-578/2/artifact/out/patch-unit-hbase-client.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-578/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-578/2/testReport/ |
| Max. process+thread count | 4978 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-578/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-client/src/main/java/org/apache/hadoop/hbase/exceptions/ClientExceptionsUtil.java [hunk 1]
```diff
@@ -87,15 +96,15 @@ public static Throwable findException(Object exception) {
     }
     Throwable cur = (Throwable) exception;
     while (cur != null) {
-      if (isSpecialException(cur)) {
+      if (isExceptionWeCare(cur)) {
```
Comment (saintstack) [inline hbase-client/src/main/java/org/apache/hadoop/hbase/exceptions/ClientExceptionsUtil.java]: Ok. Good. isSpecialException is an awful name. isExceptionWeCare is too but you made it private and it is only used here so its ok.

File: hbase-client/src/main/java/org/apache/hadoop/hbase/exceptions/ClientExceptionsUtil.java [hunk 2]
```diff
@@ -59,18 +58,28 @@ public static boolean isMetaClearingException(Throwable cur) {
...
+  }
+
+  private static boolean isInstanceOfRegionServerOverloadedException(Throwable t) {
+    return t instanceof CallQueueTooBigException || t instanceof CallDroppedException;
   }
 
+  private static boolean isMetaCachePreservingException(Throwable t) {
+    return t instanceof RegionOpeningException || t instanceof RegionTooBusyException
+        || t instanceof RpcThrottlingException || t instanceof RetryImmediatelyException
+        || t instanceof RequestTooBigException;
```
Comment (saintstack) [inline hbase-client/src/main/java/org/apache/hadoop/hbase/exceptions/ClientExceptionsUtil.java]: So, this list here is smaller than the list above in the old isSpecialException. I looked at one or two of the differences. It seems like superclasses are listed here and you drop the mention of subclasses. Is that the case? Just a little worried we're not catching something here that we used to.

File: hbase-client/src/main/java/org/apache/hadoop/hbase/exceptions/ClientExceptionsUtil.java [hunk 3]
```diff
@@ -59,18 +58,28 @@ public static boolean isMetaClearingException(Throwable cur) {
...
-        || cur instanceof RegionTooBusyException || cur instanceof RpcThrottlingException
-        || cur instanceof MultiActionResultTooLarge || cur instanceof RetryImmediatelyException
-        || cur instanceof CallQueueTooBigException || cur instanceof CallDroppedException
-        || cur instanceof NotServingRegionException || cur instanceof RequestTooBigException);
+  public static boolean isRegionServerOverloadedException(Throwable t) {
+    t = findException(t);
+    return isInstanceOfRegionServerOverloadedException(t);
+  }
+
+  private static boolean isInstanceOfRegionServerOverloadedException(Throwable t) {
```
Comment (saintstack) [inline hbase-client/src/main/java/org/apache/hadoop/hbase/exceptions/ClientExceptionsUtil.java]: Good namings.

File: hbase-client/src/main/java/org/apache/hadoop/hbase/exceptions/ClientExceptionsUtil.java (hunk 2 above)
Comment (johnhomsea) [reply hbase-client/src/main/java/org/apache/hadoop/hbase/exceptions/ClientExceptionsUtil.java]: Yes, just remove subclasses, you mean the detailed messages in the subclass?

**Comment by saintstack** [comment]
Pulled back to branch-2.2 and branch-2.1 too as well as branch-2. Doesn't look like it makes sense on master. If not, all is good else, open new issue to forward parth. Thanks @johnhomsea .

**Pull request merged**